### PR TITLE
Layer swapping proposal

### PR DESCRIPTION
**Pull request by wojciechgabrys**
I went ahead and added a layer swapping functionality, that allows users to move current layer to a new position (and put target layer in places of moved one).

Let me know if that's something you'd find useful/worth merging in – if yes, I'll add stuff like layer boundry checking and fill in translations.

layer-swapping

### CODE REVIEW

**Comment by wojciechgabrys** [comment]
(Please disregard the failed prettier check – I'll iron it out before the final commit)

**Comment by yanfali** [comment]
that's pretty awesome. However, it's not consistent with how key swapping works. So I'd prefer to make it drag and drop. I just haven't gotten around to it.

File: src/store/modules/keymap.js [hunk 1]
```diff
@@ -275,6 +275,23 @@ const mutations = {
     }
     state.layer = newLayer;
   },
+  swapLayers(state, targetLayer) {
+    if (parseInt(targetLayer) === state.layer) {
```
Comment (yanfali) [inline src/store/modules/keymap.js]: parseInt missing base

File: src/store/modules/keymap.js [hunk 2]
```diff
@@ -275,6 +275,23 @@ const mutations = {
...
+      return;
+    }
+
+    if (state.keymap[targetLayer] === undefined) {
+      mutations.initLayer(state, targetLayer);
+    }
+
+    [state.keymap[targetLayer], state.keymap[state.layer]] = [
```
Comment (yanfali) [inline src/store/modules/keymap.js]: not sure that's a good way to do it. You'd probably want to use Vue.set to make sure the reactivity system is kept in sync

**Comment by wojciechgabrys** [comment]
> that's pretty awesome. However, it's not consistent with how key swapping works. So I'd prefer to make it drag and drop. I just haven't gotten around to it.

Just to clarify, you'd like swapping to work by dragging one layer circle over another one, right? If that's the case do you have any preference regarding drag&drop library I should pull in?

And another question: Layer swap should also automatically update layer indexes on keys (like MO/TG/etc.), right?

**Comment by noroadsleft** [comment]
Will fix #188.

**Comment by yanfali** [comment]
> > that's pretty awesome. However, it's not consistent with how key swapping works. So I'd prefer to make it drag and drop. I just haven't gotten around to it.
> 
> Just to clarify, you'd like swapping to work by dragging one layer circle over another one, right? If that's the case do you have any preference regarding drag&drop library I should pull in?
> 
> And another question: Layer swap should also automatically update layer indexes on keys (like MO/TG/etc.), right?

Yes, layer updating would be necessary. We don't use a layer library, we just use native HTML5 drag and drop. Like I said, this is a lot more work than it seems.

**Comment by wojciechgabrys** [comment]
> Yes, layer updating would be necessary. We don't use a layer library, we just use native HTML5 drag and drop. Like I said, this is a lot more work than it seems.

I'll have a stab at it and see what I can do.

**Comment by yanfali** [comment]
I'll be closing this as it's no longer merge-able.

**Pull request closed without merging**